core/src/main/java/io/parsingdata/metal/data/DataExpressionSource.java

```diff
@@ -47,7 +49,7 @@ public DataExpressionSource(final ValueExpression dataExpression, final int inde
     @Override
     protected byte[] getData(final BigInteger offset, final BigInteger length) {
         final Value inputValue = getValue();
-        if (length.add(offset).compareTo(inputValue.slice.length) > 0) { throw new IllegalStateException("Data to read is not available ([offset=" + offset + ";length=" + length + ";source=" + this + ")."); }
+        if (offset.compareTo(ZERO) < 0 || length.add(offset).compareTo(inputValue.slice.length) > 0) { throw new IllegalStateException("Data to read is not available ([offset=" + offset + ";length=" + length + ";source=" + this + ")."); }
```

Comment:
Call the new `checkNotNegative` here instead?

Reply:
Same goes for the other source classes.

Reply:
That would turn it into multiple guard clauses. But I agree it's probably better...